### PR TITLE
Add random number generator resource (optionally specified with a seed)

### PR DESCRIPTION
**Pull request by LegNeato**
# Objective

- Make bevy "random aware" and allow seeds to be controlled by bevy

## Solution

- Move random number generators into bevy as a resource.

I'm new to bevy and would love guidance if this even makes sense to do. If so, I'll clean it up and add tests.

### CODE REVIEW

**Review by lafi-99 — APPROVED** [review]

**Comment by james7132** [comment]
IMO, unless these PRNGs are required for a mission-critical official crate, I'd this should be done by the developer instead of having it globally configured provided by the engine itself. There is a sizable number of open questions regarding how the PRNG is used that is entirely dependent on the developer's requirements: 
 
 - What performance characteristics is desired?
 - How random does the outputs need to be?
 - How is the PRNG seeded?
 - Does the output need to be secure?
 - Is random state shared between threads? Between isolated systems?
 - How does the PRNG's use impact determinism-sensitive systems?

These questions all have use case specific answers and there is no one-size fits all solution. Using it as a bevy_core provided resource makes the choice here for several of these questions and that might not be what's needed for every use case. The `thread_rng` cases in the examples seem to be sane enough for it's use and is a cost that only the developers who need it will pay.

**Comment by NathanSWard** [comment]
> I think this should be done by the end-developer instead of having it globally configured provided by the engine itself.

I agree. 
There are to many variables when it comes to using some sort of RNG that providing it in the engine doesn't seem to be very beneficial imo. 
Also since adding the `rand` crate to your own project and setting up an RNG is quite trivial, this is probably best left to the end-developer.

**Comment by mockersf** [comment]
While easy to setup, it's still a nice feature to have a random number generator available by default ([Godot offers one](https://docs.godotengine.org/en/stable/tutorials/math/random_number_generation.html), [Unity also](https://docs.unity3d.com/Manual/class-Random.html)).

Could you put it behind a feature flag so that it can be removed if needed?

File: crates/bevy_core/Cargo.toml [hunk 1]
```diff
@@ -15,13 +15,16 @@ keywords = ["bevy"]
...
-bevy_reflect = { path = "../bevy_reflect", version = "0.5.0", features = ["bevy"] }
-bevy_tasks = { path = "../bevy_tasks", version = "0.5.0" }
-bevy_utils = { path = "../bevy_utils", version = "0.5.0" }
+bevy_app = { path="../bevy_app", version="0.5.0" }
+bevy_derive = { path="../bevy_derive", version="0.5.0" }
+bevy_ecs = { path="../bevy_ecs", version="0.5.0" }
+bevy_math = { path="../bevy_math", version="0.5.0" }
+bevy_reflect = { path="../bevy_reflect", version="0.5.0", features=["bevy"] }
+bevy_tasks = { path="../bevy_tasks", version="0.5.0" }
+bevy_utils = { path="../bevy_utils", version="0.5.0" }
```
Comment (mockersf) [inline crates/bevy_core/Cargo.toml:24]: Could you revert those formatting change?

File: crates/bevy_core/src/rng.rs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    fn fill_bytes(&mut self, dest: &mut [u8]) {
+        self.0.fill_bytes(dest)
+    }
+    fn try_fill_bytes(&mut self, dest: &mut [u8]) -> Result<(), rand::Error> {
+        self.0.try_fill_bytes(dest)
+    }
+}
+
+#[derive(Clone)]
+pub struct InsecureRng(SmallRng);
```
Comment (mockersf) [inline crates/bevy_core/src/rng.rs:54]: Could it `impl Deref<SmallRng>`?

File: crates/bevy_core/src/rng.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        // The default is to use the same secure / cryptographic seed for both Rngs.
+        DefaultRngOptions {
+            secure_seed: seed,
+            insecure_seed: seed,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SecureRng(StdRng);
```
Comment (mockersf) [inline crates/bevy_core/src/rng.rs:31]: Could it `impl Deref<StdRng>`?

**Comment by LegNeato** [comment]
Yep, I'll clean this up and put behind a feature flag by the end of the weekend

**Comment by vgel** [comment]
Instead of being in `core`, it might make more sense for this to be a separate `bevy-random` crate / plugin that gets included as part of the `DefaultPlugins` group.

**Comment by james7132** [comment]
> I also agree it would be nice to have this by default -- little things like this add up, and while it does add a little bit of overhead for people who don't use it, I think that overhead is fairly minimal and if it's behind a feature flag it's not too difficult to remove. Having it as a plugin will also make the feature flag a bit easier to implement.

 My main concern is that this is a source of non-determinism that is fundamental and accessible enough that it may alter game state in difficult to detect ways. For games that require it, like deterministic networked games, this is can be a hard non-starter. Other game engines like Unity and Unreal that have it built in require explicit management of these global PRNGs to avoid the side effects of one system altering another.
 
Re: having it behind a feature flag: this is marginally better, though I'm concerned that other first-party crates will become dependent on it as a source of randomness, which will make it transitively required for certain core engine features.

IMO, any first party crate requiring a PRNG should use local PRNGs embedded in resources/components from the crate itself. This both keeps the dependency tree smaller and isolates the effects of one crate/system on another.

**Comment by LegNeato** [comment]
Apologies, I may not understand as I haven't worked in games professionally. The reason I put up this PR is precisely because I need bevy to be deterministic and I want anything that can change the world logic and state be driven via bevy.

IMHO a source of randomness should be a feature of the world...multiple examples have already naturally included it. Happy to make the change elsewhere / in a subcrate, but I couldn't see how to get at the very early bevy stages outside core...any pointers would be swell!

I can remove the default os randomness if that is desired. The idea was for people who care about determinism, they pass in their own seed (as in the example I added. For people who don't, they get an out of the box experience as with other engines 🤷‍♂️

**Comment by james7132** [comment]
> Apologies, I may not understand as I haven't worked in games professionally. The reason I put up this PR is precisely because I need bevy to be deterministic and I want anything that can change the world logic and state be driven via bevy.

A global PRNG means the execution order and presence of systems that depend on it will affect the current state: i.e. the particle system emitter will affect the PRNG outputs will affect the power-ups generated. It's for this reason why I think it's better to have system local PRNGs instead. Those that require control over the PRNG of multiple systems need to put in more work, but there's no spooky action at a distance by default.

**Comment by LegNeato** [comment]
Ah, I now get what you are saying!

I assumed that people would indeed add their own additional rng resources when they needed 
specifc guarantees or segmentation...but I think in the most general the distinction one would care about is insecure vs secure so those were baked in.

Isn't it also preferable to have a single source of randomness all others descend from? It helps determinism as there is one place to set the seed and all callsites respect it. As it is there is nothing preventing a system in a dependency using rand directly and causing non-deteterminism.

Perhaps instead of an rng implementation what we need as a resource is a "genesis" random seed that all rngs use to seed, something similar to  https://docs.rs/rand/0.8.4/rand/trait.SeedableRng.html#method.from_rng.

**Comment by Nilirad** [comment]
I like the idea of a seed generator. However we need to make sure that the local PRNGs are seeded deterministically.

**Comment by LegNeato** [comment]
Looking for a little more guidance on where to take this

**Comment by LegNeato** [comment]
I put up a PR with a new plugin called `bevy_entropy` in https://github.com/bevyengine/bevy/pull/2504. If that lands I'll rebase this on top of that and continue the discussion. I still think it is useful to have built-in rngs, as many of our examples use them which...likely means that many games would too. But we'll see what happens with the lower-level building block PR first.

**Comment by LegNeato** [comment]
Closing in favor of the entropy PR. Will revisit this in the future.

**Pull request closed without merging**